src/python/pants/engine/README.md

```diff
@@ -17,63 +17,74 @@ Once the engine is instantiated with a valid set of `@rule`s, a caller can synch
 computation of any of the product types provided by those `@rule`s by calling:
 
 ```python
-# Request a ThingINeed (a `Product`) for the thing_i_have (a `Subject`).
+# Request a ThingINeed (a `Product`) for a thing_i_have (a `Param`).
 thing_i_need, = scheduler.product_request(ThingINeed, [thing_i_have])
 ```
 
 The engine then takes care of concurrently executing all dependencies of the matched `@rule`s to
 produce the requested value.
 
-### Products and Subjects
+### Products and Params
 
 The engine executes your `@rule`s in order to (recursively) compute a `Product` of the requested
-type for a given `Subject`. This recursive type search leads to a very loosely coupled (and yet
+type for a set of `Param`s. This recursive type search leads to a loosely coupled (and yet
 still statically checked) form of dependency injection.
 
-When an `@rule` runs, it runs for a particular `Subject` value, which is part of the unique
-identity for that instance of the `@rule`. An `@rule` can request dependencies for different
-`Subject` values as it runs (see the section on `Get` requests below). Because the subject for
-an `@rule` is chosen by callers, a `Subject` can be of any (hashable) type that a user might want
-to compute a product for.
+When an `@rule` runs, it requires a set of `Param`s that the engine has determined are needed
+to compute its transitive `@rule` dependencies. So although an `@rule` might not have a particular
+`Param` type in its signature, it might depend on another `@rule` that does need that `Param`, and
+would thus need that `Param` in order to run. To see which `Params` the engine needs to run each
+`@rule`, refer to the `Visualization` section below.
 
-The return value of an `@rule` for a particular `Subject` is known as a `Product`. At some level,
-you can think of (`subject_value`, `product_type`) as a "key" that uniquely identifies a particular
-Product value and `@rule` execution.
+Any hashable type with useful equality may be used as a `Param`, and additional `Params` can be
+provided to an `@rule`'s dependencies via `Get` requests (see below). Each `Param` value in a set
+of `Params` is unique by type, so if `@rules` recursively introduce a particular `Param` type,
+there will still only be one value for that type in each `@rule`, but it will change as you move
+deeper into the dependency graph.
+
+The return value of an `@rule` is known as a `Product`. At some level, you can think
+of `(product_type, params_set)` as a "key" that uniquely identifies a particular `Product` value
+and `@rule` execution. If an `@rule` is able to produce a `Product` without consuming any `Params`,
```

Comment:
This sentence confused me because I thought it was closely related to the prior sentence. I recommend adding a transition word like `Further, if an @rule`.